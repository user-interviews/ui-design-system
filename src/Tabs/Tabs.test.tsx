import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { Tab, Tabs } from './index';

const tabOneKey = 'one';
const tabTwoKey = 'two';
const tabThreeKey = 'three';

const tabOneTitle = 'Tab One';
const tabTwoTitle = 'Tab Two';
const tabThreeTitle = 'Tab Three';

const tabOneContent = 'Tab Content One';
const tabTwoContent = 'Tab Content Two';
const tabThreeContent = 'Tab Content Three';

const uncontrolledProps = {
  activeKey: undefined,
  defaultActiveKey: tabOneKey,
  id: 'tabID',
  mountOnEnter: false,
  transition: false,
  unmountOnExit: false,
  onSelect: undefined,
};

const onSelectMock = jest.fn();

const controlledProps = {
  ...uncontrolledProps,
  ...{
    activeKey: tabOneKey,
    defaultActiveKey: undefined,
    onSelect: onSelectMock,
  },
};

const renderTabs = (props) => render(
  <Tabs {...props}>
    <Tab eventKey={tabOneKey} title={tabOneTitle}>
      {tabOneContent}
    </Tab>
    <Tab eventKey={tabTwoKey} title={tabTwoTitle}>
      {tabTwoContent}
    </Tab>
    <Tab disabled eventKey={tabThreeKey} title={tabThreeTitle}>
      {tabThreeContent}
    </Tab>
  </Tabs>,
);

describe('Tabs', () => {
  describe('uncontrolled tabs', () => {
    it('renders the expected snapshot', () => {
      const tabs = renderTabs(uncontrolledProps);
      expect(tabs).toMatchSnapshot();
    });

    it('shows the default tab content', () => {
      renderTabs(uncontrolledProps);
      expect(screen.getByText(tabOneContent)).toHaveClass('active');
      expect(screen.getByText(tabTwoContent)).not.toHaveClass('active');
    });

    it('shows the tab content when clicked', async () => {
      renderTabs(uncontrolledProps);

      expect(screen.getByText(tabOneContent)).toHaveClass('active');
      expect(screen.getByText(tabTwoContent)).not.toHaveClass('active');

      const tabTwoButton = screen.getByRole('tab', { name: tabTwoTitle });

      userEvent.click(tabTwoButton);
      await waitFor(() => {
        expect(tabTwoButton).toHaveClass('active');
      });

      expect(screen.getByText(tabOneContent)).not.toHaveClass('active');
      expect(screen.getByText(tabTwoContent)).toHaveClass('active');
    });
  });

  describe('controlled tabs', () => {
    it('renders the expected snapshot', () => {
      const tabs = renderTabs(controlledProps);
      expect(tabs).toMatchSnapshot();
    });

    it('shows the default tab content', () => {
      renderTabs(controlledProps);
      expect(screen.getByText(tabOneContent)).toHaveClass('active');
      expect(screen.getByText(tabTwoContent)).not.toHaveClass('active');
    });

    it('shows the tab content when clicked', async () => {
      renderTabs(controlledProps);

      expect(screen.getByText(tabOneContent)).toHaveClass('active');
      expect(screen.getByText(tabTwoContent)).not.toHaveClass('active');

      const tabTwoButton = screen.getByRole('tab', { name: tabTwoTitle });

      userEvent.click(tabTwoButton);
      await waitFor(() => {
        expect(onSelectMock).toHaveBeenCalled();
      });
    });
  });
});
