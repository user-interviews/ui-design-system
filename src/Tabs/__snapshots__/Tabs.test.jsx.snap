// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs controlled tabs renders the expected snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <span
        class="sc-bdvvtL eGviLc"
      >
        <ul
          class="nav nav-tabs"
          role="tablist"
        >
          <li
            class="nav-item"
            role="presentation"
          >
            <button
              aria-controls="tabID-tabpane-one"
              aria-selected="true"
              class="nav-link active"
              data-rr-ui-event-key="one"
              id="tabID-tab-one"
              role="tab"
              type="button"
            >
              Tab One
            </button>
          </li>
          <li
            class="nav-item"
            role="presentation"
          >
            <button
              aria-controls="tabID-tabpane-two"
              class="nav-link"
              data-rr-ui-event-key="two"
              id="tabID-tab-two"
              role="tab"
              tabindex="-1"
              type="button"
            >
              Tab Two
            </button>
          </li>
          <li
            class="nav-item"
            role="presentation"
          >
            <button
              aria-controls="tabID-tabpane-three"
              aria-disabled="true"
              class="nav-link disabled"
              data-rr-ui-event-key="three"
              disabled=""
              id="tabID-tab-three"
              role="tab"
              tabindex="-1"
              type="button"
            >
              Tab Three
            </button>
          </li>
        </ul>
        <div
          class="tab-content"
        >
          <div
            aria-labelledby="tabID-tab-one"
            class="tab-pane active"
            id="tabID-tabpane-one"
          >
            Tab Content One
          </div>
          <div
            aria-labelledby="tabID-tab-two"
            class="tab-pane"
            id="tabID-tabpane-two"
          >
            Tab Content Two
          </div>
          <div
            aria-labelledby="tabID-tab-three"
            class="tab-pane"
            id="tabID-tabpane-three"
          >
            Tab Content Three
          </div>
        </div>
      </span>
    </div>
  </body>,
  "container": <div>
    <span
      class="sc-bdvvtL eGviLc"
    >
      <ul
        class="nav nav-tabs"
        role="tablist"
      >
        <li
          class="nav-item"
          role="presentation"
        >
          <button
            aria-controls="tabID-tabpane-one"
            aria-selected="true"
            class="nav-link active"
            data-rr-ui-event-key="one"
            id="tabID-tab-one"
            role="tab"
            type="button"
          >
            Tab One
          </button>
        </li>
        <li
          class="nav-item"
          role="presentation"
        >
          <button
            aria-controls="tabID-tabpane-two"
            class="nav-link"
            data-rr-ui-event-key="two"
            id="tabID-tab-two"
            role="tab"
            tabindex="-1"
            type="button"
          >
            Tab Two
          </button>
        </li>
        <li
          class="nav-item"
          role="presentation"
        >
          <button
            aria-controls="tabID-tabpane-three"
            aria-disabled="true"
            class="nav-link disabled"
            data-rr-ui-event-key="three"
            disabled=""
            id="tabID-tab-three"
            role="tab"
            tabindex="-1"
            type="button"
          >
            Tab Three
          </button>
        </li>
      </ul>
      <div
        class="tab-content"
      >
        <div
          aria-labelledby="tabID-tab-one"
          class="tab-pane active"
          id="tabID-tabpane-one"
        >
          Tab Content One
        </div>
        <div
          aria-labelledby="tabID-tab-two"
          class="tab-pane"
          id="tabID-tabpane-two"
        >
          Tab Content Two
        </div>
        <div
          aria-labelledby="tabID-tab-three"
          class="tab-pane"
          id="tabID-tabpane-three"
        >
          Tab Content Three
        </div>
      </div>
    </span>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Tabs uncontrolled tabs renders the expected snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <span
        class="sc-bdvvtL eGviLc"
      >
        <ul
          class="nav nav-tabs"
          role="tablist"
        >
          <li
            class="nav-item"
            role="presentation"
          >
            <button
              aria-controls="tabID-tabpane-one"
              aria-selected="true"
              class="nav-link active"
              data-rr-ui-event-key="one"
              id="tabID-tab-one"
              role="tab"
              type="button"
            >
              Tab One
            </button>
          </li>
          <li
            class="nav-item"
            role="presentation"
          >
            <button
              aria-controls="tabID-tabpane-two"
              class="nav-link"
              data-rr-ui-event-key="two"
              id="tabID-tab-two"
              role="tab"
              tabindex="-1"
              type="button"
            >
              Tab Two
            </button>
          </li>
          <li
            class="nav-item"
            role="presentation"
          >
            <button
              aria-controls="tabID-tabpane-three"
              aria-disabled="true"
              class="nav-link disabled"
              data-rr-ui-event-key="three"
              disabled=""
              id="tabID-tab-three"
              role="tab"
              tabindex="-1"
              type="button"
            >
              Tab Three
            </button>
          </li>
        </ul>
        <div
          class="tab-content"
        >
          <div
            aria-labelledby="tabID-tab-one"
            class="tab-pane active"
            id="tabID-tabpane-one"
          >
            Tab Content One
          </div>
          <div
            aria-labelledby="tabID-tab-two"
            class="tab-pane"
            id="tabID-tabpane-two"
          >
            Tab Content Two
          </div>
          <div
            aria-labelledby="tabID-tab-three"
            class="tab-pane"
            id="tabID-tabpane-three"
          >
            Tab Content Three
          </div>
        </div>
      </span>
    </div>
  </body>,
  "container": <div>
    <span
      class="sc-bdvvtL eGviLc"
    >
      <ul
        class="nav nav-tabs"
        role="tablist"
      >
        <li
          class="nav-item"
          role="presentation"
        >
          <button
            aria-controls="tabID-tabpane-one"
            aria-selected="true"
            class="nav-link active"
            data-rr-ui-event-key="one"
            id="tabID-tab-one"
            role="tab"
            type="button"
          >
            Tab One
          </button>
        </li>
        <li
          class="nav-item"
          role="presentation"
        >
          <button
            aria-controls="tabID-tabpane-two"
            class="nav-link"
            data-rr-ui-event-key="two"
            id="tabID-tab-two"
            role="tab"
            tabindex="-1"
            type="button"
          >
            Tab Two
          </button>
        </li>
        <li
          class="nav-item"
          role="presentation"
        >
          <button
            aria-controls="tabID-tabpane-three"
            aria-disabled="true"
            class="nav-link disabled"
            data-rr-ui-event-key="three"
            disabled=""
            id="tabID-tab-three"
            role="tab"
            tabindex="-1"
            type="button"
          >
            Tab Three
          </button>
        </li>
      </ul>
      <div
        class="tab-content"
      >
        <div
          aria-labelledby="tabID-tab-one"
          class="tab-pane active"
          id="tabID-tabpane-one"
        >
          Tab Content One
        </div>
        <div
          aria-labelledby="tabID-tab-two"
          class="tab-pane"
          id="tabID-tabpane-two"
        >
          Tab Content Two
        </div>
        <div
          aria-labelledby="tabID-tab-three"
          class="tab-pane"
          id="tabID-tabpane-three"
        >
          Tab Content Three
        </div>
      </div>
    </span>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
